### PR TITLE
Add determinant vertex

### PR DESCRIPTION
**Pull request by Pinpickle**
This PR adds support for a determinant vertex. It takes a vertex shaped like a matrix as input, and outputs a scalar determinant.

Reverse auto diff has been implemented on this vertex as well. It's slightly problematic as it involves the inverse of our input matrix, which fails when the determinant is 0 (despite the gradient being well-defined at that point). This seems to be the expected way to perform such a calculation, and none of the literature even addresses this as a problem. @gordoncaleb and I couldn't think of a case where this would be a problem, and I added some random testing to verify this at least in a small case.

Forward differentiation has not been implemented.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Map<Vertex, PartialDerivatives> reverseModeAutoDifferentiation(PartialDerivatives derivativeOfOutputsWithRespectToSelf) {
+        // As this is the inverse, this will fail if the matrixDeterminant is zero
+        DoubleTensor inverseTranspose = inputVertex.getValue().transpose().matrixInverse();
+
+        // Reverse mode auto diff of matrix matrixDeterminant: https://www.cs.ox.ac.uk/files/723/NA-08-01.pdf
+        // Using variables from the paper
```
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Looks useful, might be better to put it in a javadoc for the method?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/DoubleVertex.java [hunk 2]
```diff
@@ -75,6 +76,17 @@ public DoubleVertex matrixInverse() {
         return new MatrixInverseVertex(this);
     }
 
+    /**
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/DoubleVertex.java]: The method name here should be enough to describe what's going on. I'd just leave the javadoc for the class itself.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Method javadocs are better than inline comments. With javadoc a user can see the comments in the docs instead of having to dig through the source.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    protected DualNumber dualOp(DualNumber dualNumber) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Map<Vertex, PartialDerivatives> reverseModeAutoDifferentiation(PartialDerivatives derivativeOfOutputsWithRespectToSelf) {
+        // As this is the inverse, this will fail if the matrixDeterminant is zero
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: This comment should be on the matrixInverse() method instead of here I think. You could argue that that's a property of the function instead of a quirk of the code that needs documenting.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    @Test
+    public void canOptimiseOutOfTheBoxEvenWithRiskOfSingularMatrices() {
+        // The standard method for calculating the gradient of the determinant fails when the matrixDeterminant is 0.
+        // Regardless, the gradient is defined when the matrixDeterminant is 0.
+        // We run the optimizer many times to create some confidence that we won't run into problems in a real-world
+        // use-case.
+        final int runCount = 100;
+        IntStream.range(0, runCount).forEach(i -> {
+            int[] shape = new int[]{2, 2};
+            DoubleVertex input = new GaussianVertex(shape, Math.random() * 2 - 4, Math.random() * 2 + 0.1);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: Nondeterministic tests are really a pain. Think about the edge cases this random test might find and then test them explicitly and deterministically.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Test
+    public void canDifferentiateWrtTensor() {
+        final int[] shape = new int[]{2, 2};
+        final DoubleVertex input = new UniformVertex(shape, 0, 10);
+        input.setValue(DoubleTensor.create(new double[]{1, 2, 3, 4}, shape));
+        final DoubleVertex output = input.matrixDeterminant().times(input);
+
+        // For calculation of derivative, see
+        // https://www.wolframalpha.com/input/?i=x+%3D+1;+y+%3D+2;+z+%3D+3;+w+%3D+4;+d%2Fdx+(%7B%7Bx,y%7D,%7Bz,w%7D%7D+*+det(%7B%7Bx,y%7D,%7Bz,w%7D%7D))
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: Awesome work with the wolframalpha link!

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java (hunk 4 above)
Comment (Pinpickle) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: That's a good point. This loop is also making the test take much longer than it needs to.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    protected DoubleTensor op(DoubleTensor value) {
+        return DoubleTensor.scalar(value.determinant());
+    }
+
+    @Override
+    protected DualNumber dualOp(DualNumber dualNumber) {
+        throw new UnsupportedOperationException();
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Potentially a few reasons to implement this:
1) To provide extra confidence to our gradient tests (we can check that forward = reverse etc)
2) Given we have a limitation in reverse - might be best if people can still get the gradient if they hit the 0 case...
3) There will be certain problems where forward accumulation is quicker.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
+
+/**
+ * A vertex that takes on the value of the matrixDeterminant of the value of its input matrix
+ * <p>
+ * Gradient calculations (and thus gradient-based optimisations) will fail if the matrixDeterminant is 0.
+ */
+public class MatrixDeterminantVertex extends DoubleUnaryOpVertex {
+    public MatrixDeterminantVertex(DoubleVertex vertex) {
```
Comment (GeorgeNash) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: We've only implemented determinant for rank 2 tensors right? Let's put in an check to ensure `vertex` is a matrix and throw an `IllegalArgumentException` if not

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    public void canOptimiseOutOfTheBoxStartingAtZero() {
+        assertOptimizerWorks(0);
+    }
+
+    private void assertReverseAutoDiffMatches(DoubleVertex output, DoubleVertex input, DoubleTensor expectedDerivative) {
+        final DoubleTensor derivative = Differentiator.reverseModeAutoDiff(output, input).withRespectTo(input);
+        Assert.assertThat(derivative, TensorMatchers.elementwiseEqualTo(expectedDerivative));
+    }
+
+    private void assertOptimizerWorks(double inputGaussianMu) {
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: This doesn't really mean anything - can we call it `assertOptimizerWorksWithDeterminant`?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            Optimizer.of(net).maxLikelihood();
+            Assert.assertEquals(input.getValue().determinant(), 2.2, 0.1);
+        });
+    }
+
+    private void assertReverseAutoDiffMatches(DoubleVertex output, DoubleVertex input, DoubleTensor expectedDerivative) {
+        final DoubleTensor derivative = Differentiator.reverseModeAutoDiff(output, input).withRespectTo(input);
+        Assert.assertThat(derivative, TensorMatchers.elementwiseEqualTo(expectedDerivative));
+    }
+}
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: `finiteDifferenceMatchesGradient`

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.tensor.TensorMatchers;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.dbl.Differentiator;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.UniformVertex;
+
+public class MatrixDeterminantVertexTest {
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: Nice tests! In all your tests the `determinant` is never the end of the graph and is never the thing we are calculating the gradient `of`. Can we add in a test where the determinant is at the end?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
+
+/**
+ * A vertex that takes on the value of the matrixDeterminant of the value of its input matrix
+ * <p>
+ * Gradient calculations (and thus gradient-based optimisations) will fail if the matrixDeterminant is 0.
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Can you demonstrate this in a test?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    @Test
+    public void calculatesDeterminant() {
+        final DoubleVertex input = new ConstantDoubleVertex(DoubleTensor.create(new double[]{1, 2, 3, 4}, 2, 2));
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: It's probably worth having a test that uses a matrix larger than 2x2

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
```
[the remark this answers is not in the document]
Comment (Pinpickle) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: I think that makes reading this harder (in the case of 4 items), as we have to call `.arange(1,5).reshape(shape)`. The exclusivity of the range in particular makes this harder to read.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 6 above)
Comment (Pinpickle) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: This was a good idea and @gordoncaleb and I spent some time investigating this. At this stage it's too much effort to implement. I'll make this clearer.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: That's the badger

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 6 above)
Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Cool - thanks.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Pinpickle) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: This is muuuch nicer

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 7 above)
Comment (Pinpickle) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Yes good catch!

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 11 above)
Comment (Pinpickle) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Went back and forth on whether to do this a few times. In the end I decided to. If it's in our javadoc, it's part of the API contract, so it should be tested for. Unfortunately due to the way forward mode differentiation is implemented, I can't test for that not being implemented (I don't think)

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    public MatrixDeterminantVertex(DoubleVertex vertex) {
+        super(Tensor.SCALAR_SHAPE, vertex);
+        assertInputValueIsSquareMatrix(vertex.getValue());
+    }
+
+    private static void assertInputValueIsSquareMatrix(DoubleTensor value) {
+        final IllegalArgumentException exception = new IllegalArgumentException("Input tensor must be a square matrix");
+
+        if (value == null) {
+            throw exception;
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: `Preconditions.checkNotNull(...)`

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        final int[] shape = value.getShape();
+
+        if ((shape.length != 2) || (shape[0] != shape[1])) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Maybe use two booleans `isNotMatrix` and `isNotSquare` and clean this predicate up.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+ * Forward mode differentiation is not implemented due to requiring a tensor trace, which is not yet implemented
+ */
+public class MatrixDeterminantVertex extends DoubleUnaryOpVertex {
+    public MatrixDeterminantVertex(DoubleVertex vertex) {
+        super(Tensor.SCALAR_SHAPE, vertex);
+        assertInputValueIsSquareMatrix(vertex.getValue());
+    }
+
+    private static void assertInputValueIsSquareMatrix(DoubleTensor value) {
+        final IllegalArgumentException exception = new IllegalArgumentException("Input tensor must be a square matrix");
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: No need to create this on every call to this assert...

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 14 above)
Comment (Pinpickle) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Going to drop this as we discussed as null checking isn't something we really do in the codebase

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        Assert.assertThat(input.matrixDeterminant().getValue(), TensorMatchers.isScalarWithValue(-2d));
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: You can drop the `TensorMatchers.` for better readability. 

i.e. then you will have
`import io.improbable.keanu.tensor.TensorMatchers.isScalarWithValue`
instead of
`import io.improbable.keanu.tensor.TensorMatchers`

Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: likewise you can drop the `Assert.`

Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: You can use `VertexMatchers#hasValue` instead:
`assertThat(input.matrixDeterminant(), hasValue(-2d));`

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    private static void assertInputValueIsSquareMatrix(DoubleTensor value) {
+        final int[] shape = value.getShape();
+        final boolean isMatrix = shape.length == 2;
+        final boolean isSquare = isMatrix && shape[0] == shape[1];
+
+        if (!isMatrix || !isSquare) {
+            throw new IllegalArgumentException("Input tensor must be a square matrix");
+        }
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: It would be better to separate out the 2 failure conditions into separate error messages. So you end up with:

    if (shape.length != 2) {
            throw new IllegalArgumentException("Input tensor must be a matrix");
    } else if (shape[0] != shape[1]) {
            throw new IllegalArgumentException("Input matrix must be square");
    }

Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: The natural place for this method to live is `TensorShapeValidation`

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    @Test
+    public void calculatesDeterminantOnLargeMatrix() {
+        final double[] values = new double[]{
+            1, 8, 4, 5, 6,
+            1, 100, 232, 4, 15,
+            -10, 3, 57, 68, 10,
+            7, 8, 9, 20, 10,
+            12, 32, 43, -2, 5};
+        final DoubleVertex input = new ConstantDoubleVertex(DoubleTensor.create(values, 5, 5));
+        Assert.assertThat(input.matrixDeterminant().getValue(), TensorMatchers.isScalarWithValue(Matchers.closeTo(6120880d, 1e-5)));
```
[the remark this answers is not in the document]
Comment (Pinpickle) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java]: Determinant calculations end up with big numbers for larger matrices - this calculation in particular ended up being not exact (so an exact match would fail)

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java (hunk 18 above)
Comment (Pinpickle) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertex.java]: Good call - moved and added tests
